synapse/storage/databases/main/stats.py

```diff
@@ -242,7 +243,7 @@ async def update_room_state(self, room_id: str, fields: Dict[str, Any]) -> None:
             "canonical_alias",
         ):
             field = fields.get(col)
-            if field and "\0" in field:
+            if not isinstance(field, str) or "\0" in field:
```

Comment:
I'm not sure if it matters, but this changes the behaviour of fields that do not exist or are NULL.

Before, they would not be part of an upsert — now they will be upserted to `NULL`.

Is this fine? I get a funny feeling it may not be.

Reply:
Yes, I think this would violate the assumption that seems to be part of https://github.com/matrix-org/synapse/blob/b5133dd97f693ca213b30f4f3e874e9ab3958ea7/synapse/handlers/stats.py#L297 ; I believe it will give you a dict only populated with the fields that have changed.

I must apologise for the lack of docstring which makes it clear, though it also looks like I inherited this function..?

Reply:
Ah, I was assuming that the dictionary would include strings or None for all fields, but you're 100% correct that this does slightly change the behavior!

In the case of invalid data, does upserting with `None` make sense?

Reply:
The code as-is will never "clear" a value (e.g. if topic goes from "foo" -> None it will be stale and stay at "foo" forever). I'm unsure if this is correct behavior.

Reply:
> The code as-is will never "clear" a value (e.g. if topic goes from "foo" -> None it will be stale and stay at "foo" forever). I'm unsure if this is correct behavior.

I guess I didn't think too hard about unsetting topics; can you do that, or would it be empty string? (not that empty string was properly handled to begin with, by the looks)...

Reply:
I think upserting with None when you get invalid data is fine, if that corresponds with the interpretation that clients, etc are meant to have.

If you get a history_visibility entry with an invalid value, what happens? Stays the same, or resets to some 'default' value? I suppose that is ultimately what might guide us here.

(Though this isn't used for any important information other than the room directory really, so as long as it always leads to a 'safer' outcome I suppose it's fine; e.g. not revealing private rooms in the public room list even if they have sent nonsense visibility and join rule state events)...

Reply:
I pushed some changes which should stop overriding data that wasn't submitted. I also fixed the issue of being unable to submit `None` values into the method. I'm not 100% sure that's correct or not though.